### PR TITLE
1584: Require re-review of a PR if the target branch changes

### PR DESCRIPTION
**Pull request by erikj79**
This change improves reviewer handling when the target branch of a PR is changed. Currently this only invalidates a review on GitHub, but not on GitLab. With this change the behavior is made uniform. In addition to this, the following new behavior is added.

1. When the target branch of a PR is changed, any existing reviews in the Reviewers list will be flagged as "Re-review required (review applies to pull request targeting <branch>)" to make what happened clearer.
2. If the target branch of a PR is changed back to a previously targeted branch, any reviews made against that branch may now become valid again (if all other requirements are met).

Note that changes to and from pr/X branches are ignored, just like before.

PullRequest::reviews is no longer specifying the order of reviews returned. This order was actually not consistent between implementations and I would rather we use explicit sorts when needed instead of assuming an order.

I'm not really happy with the placement of the new static method `calculateReviewTargetRefs`, but I couldn't really find a better location, and I really wanted to share this between different `PullRequest` implementations. I think the correct solution would be to introduce an `AbstractPullRequest` common super class, but I didn't want to do this for a single method.

I have verified this with the new and modified tests, as well as by running against the playground repo(s).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [SKARA-1584](https://bugs.openjdk.org/browse/SKARA-1584): Require re-review of a PR if the target branch changes

### Reviewers
 * [Erik Helin](https://openjdk.org/census#ehelin) (@edvbld - **Reviewer**) ⚠️ Review applies to [afc188de](https://git.openjdk.org/skara/pull/1383/files/afc188de39ee9afe494b73b1a0e28feef8e21884)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara pull/1383/head:pull/1383` \
`$ git checkout pull/1383`

Update a local copy of the PR: \
`$ git checkout pull/1383` \
`$ git pull https://git.openjdk.org/skara pull/1383/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1383`

View PR using the GUI difftool: \
`$ git pr show -t 1383`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1383.diff">https://git.openjdk.org/skara/pull/1383.diff</a>

</details>

### CODE REVIEW

**Comment by erikj79** [comment]
I believe the tests are failing due to intermittent failures in the poller. That should be fixed by https://git.openjdk.org/skara/pull/1380.

File: forge/src/main/java/org/openjdk/skara/forge/RefChange.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package org.openjdk.skara.forge;
+
+import java.time.ZonedDateTime;
+
+public record RefChange(String prevRefName, String curRefName, ZonedDateTime createdAt) {
```
Comment (edvbld) [inline forge/src/main/java/org/openjdk/skara/forge/RefChange.java:5]: Again, I'm more for shorter names, so I would have gone with `ReferenceChange(String from, String to, ZonedDateTime at)`

**Review by edvbld — APPROVED** [review]
This patch looks good, just some stylistic nits.

**Comment by edvbld** [comment]
I don't really mind the static method, in general I prefer that over an abstract base class. I personally think that an abstract base class makes it so much harder to follow the control flow in sub-classes, whereas now it is very clear in `GitLabMergeRequest` and `GitHubPullRequest` how the static helper method is used.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java [hunk 2]
```diff
@@ -500,19 +500,24 @@ private Optional<String> getReviewersList() {
...
-                                       if (!hash.get().equals(pr.headHash())) {
-                                           if (ignoreStaleReviews) {
-                                               entry += " 🔄 Re-review required (review applies to [" + hash.get().abbreviate()
-                                                     + "](" + pr.filesUrl(hash.get()) + "))";
-                                           } else {
-                                               entry += " ⚠️ Review applies to [" + hash.get().abbreviate()
-                                                       + "](" + pr.filesUrl(hash.get()) + ")";
+                                   if (!review.targetRef().equals(pr.targetRef())) {
+                                       entry += " 🔄 Re-review required (review applies to pull request targeting [" + review.targetRef()
+                                               + "](" + pr.repository().webUrl(new Branch(review.targetRef())) + "))";
```
[the remark this answers is not in the document]
Comment (erikj79) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:505]: Yes, that reads better, but I will keep the link.

File: forge/src/main/java/org/openjdk/skara/forge/Review.java [hunk 3]
```diff
@@ -36,14 +36,27 @@ public class Review {
...
         this.createdAt = createdAt;
         this.reviewer = reviewer;
         this.verdict = verdict;
         this.hash = hash;
         this.id = id;
         this.body = body;
+        this.targetRef = targetRef;
+    }
+
+    public Review(Review other, String targetRef) {
```
[the remark this answers is not in the document]
Comment (erikj79) [reply forge/src/main/java/org/openjdk/skara/forge/Review.java:52]: I like the instance method.

File: forge/src/main/java/org/openjdk/skara/forge/RefChange.java (hunk 1 above)
Comment (erikj79) [reply forge/src/main/java/org/openjdk/skara/forge/RefChange.java:5]: I like that.

**Pull request closed without merging**